sw/host/opentitanlib/src/test_utils/lc_transition.rs

```diff
@@ -0,0 +1,146 @@
+    MutexAlreadyClaimed,
+    #[error("Failed to claim LC transition mutex.")]
+    FailedToClaimMutex,
+    #[error("LC transition target programming failed (target state: {0:x}).")]
+    TargetProgrammingFailed(u32),
+    #[error("LC transition failed.")]
+    TransitionFailed,
+    #[error("Bad post transition LC state: {0:x}.")]
+    BadPostTransitionState(u32),
+    #[error("Generic error {0}")]
+    Generic(String),
+}
+impl_serializable_error!(LcTransitionError);
+
+pub fn trigger_lc_transition(
+    transport: &TransportWrapper,
+    jtag: Rc<dyn Jtag>,
+    target_lc_state: DifLcCtrlState,
+    use_external_clk: bool,
+    reset_delay: Duration,
+) -> Result<()> {
+    const MAX_LC_TRANSITIONS: u32 = 24;
+
+    // Check the lc_ctrl is initialized and ready to accept a transition request.
+    let lc_status = jtag.read_lc_ctrl_reg(&LcCtrlReg::Status)?;
+    let expected_lc_status =
+        LcCtrlStatusBit::union([LcCtrlStatusBit::Initialized, LcCtrlStatusBit::Ready]);
+    if lc_status != expected_lc_status {
+        return Err(LcTransitionError::LcCtrlNotReady(lc_status).into());
+    }
+
+    // Check we are under the maximum LC transition count thresholds.
+    let lc_transition_cnt = jtag.read_lc_ctrl_reg(&LcCtrlReg::LcTransitionCnt)?;
+    if lc_transition_cnt >= MAX_LC_TRANSITIONS {
+        return Err(LcTransitionError::MaxTransitionCountExceeded(lc_transition_cnt).into());
+    }
+
+    // Check the LC transition mutex has not been claimed yet.
+    if jtag.read_lc_ctrl_reg(&LcCtrlReg::ClaimTransitionIf)? == u8::from(MultiBitBool8::True) as u32
+    {
+        return Err(LcTransitionError::MutexAlreadyClaimed.into());
+    }
+
+    // Attempt to claim the LC transition mutex.
+    jtag.write_lc_ctrl_reg(
+        &LcCtrlReg::ClaimTransitionIf,
+        u8::from(MultiBitBool8::True) as u32,
+    )?;
+
+    // Check the LC transition mutex was claimed.
+    if jtag.read_lc_ctrl_reg(&LcCtrlReg::ClaimTransitionIf)? != u8::from(MultiBitBool8::True) as u32
+    {
+        return Err(LcTransitionError::FailedToClaimMutex.into());
+    }
+
+    // Program the target LC state, i.e., Scrap.
+    jtag.write_lc_ctrl_reg(
+        &LcCtrlReg::TransitionTarget,
+        target_lc_state.redundant_encoding(),
+    )?;
+
+    // Check correct target LC state was programmed.
+    let target_lc_state_programmed = jtag.read_lc_ctrl_reg(&LcCtrlReg::TransitionTarget)?;
+    if target_lc_state_programmed != target_lc_state.redundant_encoding() {
+        return Err(LcTransitionError::TargetProgrammingFailed(target_lc_state_programmed).into());
+    }
+
+    // Configure external clock.
+    if use_external_clk {
+        jtag.write_lc_ctrl_reg(
+            &LcCtrlReg::TransitionCtrl,
+            LcCtrlTransitionCtrlBit::union([LcCtrlTransitionCtrlBit::ExtClockEn]),
+        )?;
+    } else {
+        jtag.write_lc_ctrl_reg(&LcCtrlReg::TransitionCtrl, 0)?;
+    }
+
+    // Check external clock was not requested.
+    assert_ne!(
```

Comment:
What is the purpose of this assertion? If the external clock was requested then this is going to fail.